### ISSUE TITLE
valgrind noise...

### ISSUE THREAD

**Issue by usergoodvery** [curl/curl#1117]
### I did this
#0  0x00000000059e44d5 in BN_bin2bn (s=0x139ec030 "", len=32, ret=0x139ebf50)
    at bn_lib.c:638
#1  0x00000000059ea8a1 in bnrand (pseudorand=0, rnd=0x139ebf50, bits=256,
    top=-1, bottom=0) at bn_rand.c:192
#2  0x00000000059ea973 in BN_rand (rnd=0x139ebf50, bits=256, top=-1, bottom=0)
    at bn_rand.c:206
#3  0x00000000059eabbe in bn_rand_range (pseudo=0, r=0x139ebf50,
    range=0x139ebe70) at bn_rand.c:272
#4  0x00000000059eac9d in BN_rand_range (r=0x139ebf50, range=0x139ebe70)
    at bn_rand.c:289
#5  0x0000000005a11a58 in EC_KEY_generate_key (eckey=0x139eb530)
    at ec_key.c:263
#6  0x00000000056a4eff in ssl3_send_client_key_exchange (s=0x139b1d70)
    at s3_clnt.c:2869
#7  0x000000000569f6e0 in ssl3_connect (s=0x139b1d70) at s3_clnt.c:430
#8  0x00000000056cd695 in SSL_connect (s=0x139b1d70) at ssl_lib.c:1008
#9  0x00000000056b5132 in ssl23_get_server_hello (s=0x139b1d70)
    at s23_clnt.c:799
#10 0x00000000056b3fa9 in ssl23_connect (s=0x139b1d70) at s23_clnt.c:228
#11 0x00000000056cd695 in SSL_connect (s=0x139b1d70) at ssl_lib.c:1008
#12 0x0000000006023ff4 in ossl_connect_step2 (conn=0x134e2498, sockindex=0)
    at vtls/openssl.c:2138
#13 0x0000000006025e90 in ossl_connect_common (conn=0x134e2498, sockindex=0,
    nonblocking=true, done=0x117a8144) at vtls/openssl.c:2974
#14 0x0000000006025fc2 in Curl_ossl_connect_nonblocking (conn=0x134e2498,
    sockindex=0, done=0x117a8144) at vtls/openssl.c:3008
#15 0x0000000006026e93 in Curl_ssl_connect_nonblocking (conn=0x134e2498,
    sockindex=0, done=0x117a8144) at vtls/vtls.c:322
#16 0x0000000005fbf15d in https_connecting (conn=0x134e2498, done=0x117a8144)
    at http.c:1385
#17 0x0000000005fbf0bb in Curl_http_connect (conn=0x134e2498, done=0x117a8144)
    at http.c:1355
#18 0x0000000005fd87d9 in Curl_protocol_connect (conn=0x134e2498,
    protocol_done=0x117a8144) at url.c:3727
#19 0x0000000005ff48a1 in multi_runsingle (multi=0x134ba6f8, now=...,
    data=0xf41abe8) at multi.c:1567
#20 0x0000000005ff5a7a in curl_multi_perform (multi=0x134ba6f8,
    running_handles=0x117a830c) at multi.c:2115
#21 0x0000000005fe9f6a in easy_transfer (multi=0x134ba6f8) at easy.c:727
#22 0x0000000005fea14c in easy_perform (data=0xf41abe8, events=false)
    at easy.c:814
#23 0x0000000005fea196 in curl_easy_perform (data=0xf41abe8) at easy.c:833
#24 0x0000000000432c39 in HttpRequestGoogleGcm (http_ptr=0x117a8cb0,
    url_str=0x4a9e00 "https://android.googleapis.com/gcm/send",
    json_payload=0x134e1ff0 "{\"registration_ids\":[\"valuex\"] \"message\":\"h\"}") at http_request.c:177
### curl/libcurl version
libssl.so.1.0.0 => /opt/lib/libssl.so.1.0.0 (0x00007f18f0f20000)
libcrypto.so.1.0.0 => /opt/lib/libcrypto.so.1.0.0 (0x00007f18f0a8e000)
 libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f18f0878000)
  libcurl.so.4 => /opt/lib/libcurl.so.4 (0x00007f18f05c6000)

[curl -V output perhaps?]

### operating system
debian Linux 4.4.0-x86_64-linode63 x86_64

I'm getting some valgrind instrumentation noise when I connect using libcurl. always found it tricky to rationalise about some of the uninitialised warnings spat out by valgrind, so it maybe  a false one.

==00:00:00:31.910 20121== Conditional jump or move depends on uninitialised value(s)
==00:00:00:31.911 20121==    at 0x59E44D5: BN_bin2bn (bn_lib.c:638)
==00:00:00:31.911 20121==    by 0x59EA8A0: bnrand (bn_rand.c:192)
==00:00:00:31.911 20121==    by 0x59EA972: BN_rand (bn_rand.c:206)
==00:00:00:31.911 20121==    by 0x59EABBD: bn_rand_range (bn_rand.c:272)
==00:00:00:31.911 20121==    by 0x59EAC9C: BN_rand_range (bn_rand.c:289)
==00:00:00:31.911 20121==    by 0x5A11A57: EC_KEY_generate_key (ec_key.c:263)
==00:00:00:31.911 20121==    by 0x56A4EFE: ssl3_send_client_key_exchange (s3_clnt.c:2869)
==00:00:00:31.911 20121==    by 0x569F6DF: ssl3_connect (s3_clnt.c:430)
==00:00:00:31.911 20121==    by 0x56CD694: SSL_connect (ssl_lib.c:1008)
==00:00:00:31.911 20121==    by 0x56B5131: ssl23_get_server_hello (s23_clnt.c:799)
==00:00:00:31.911 20121==    by 0x56B3FA8: ssl23_connect (s23_clnt.c:228)
==00:00:00:31.911 20121==    by 0x56CD694: SSL_connect (ssl_lib.c:1008)
==00:00:00:31.911 20121==  Uninitialised value was created by a heap allocation
==00:00:00:31.911 20121==    at 0x4C295E4: malloc (vg_replace_malloc.c:299)
==00:00:00:31.911 20121==    by 0x5972B0E: default_malloc_ex (mem.c:79)
==00:00:00:31.911 20121==    by 0x59731C5: CRYPTO_malloc (mem.c:342)
==00:00:00:31.911 20121==    by 0x59EA69A: bnrand (bn_rand.c:138)
==00:00:00:31.911 20121==    by 0x59EA972: BN_rand (bn_rand.c:206)
==00:00:00:31.911 20121==    by 0x59EABBD: bn_rand_range (bn_rand.c:272)
==00:00:00:31.911 20121==    by 0x59EAC9C: BN_rand_range (bn_rand.c:289)
==00:00:00:31.911 20121==    by 0x5A11A57: EC_KEY_generate_key (ec_key.c:263)
==00:00:00:31.911 20121==    by 0x56A4EFE: ssl3_send_client_key_exchange (s3_clnt.c:2869)
==00:00:00:31.911 20121==    by 0x569F6DF: ssl3_connect (s3_clnt.c:430)
==00:00:00:31.911 20121==    by 0x56CD694: SSL_connect (ssl_lib.c:1008)
==00:00:00:31.911 20121==    by 0x56B5131: ssl23_get_server_hello (s23_clnt.c:799)

**Comment by bagder** [comment]
Sorry, but that's an OpenSSL issue mentioned in their FAQ: [Why does Valgrind complain about the use of uninitialized data?](https://www.openssl.org/docs/faq.html#PROG14)

They have addressed that in the 1.1.0 series.

**Comment by phmccarty** [comment]
The FAQ no longer contains that entry on Valgrind, so I'll paste the text I found in the openssl git history for the record.

```
* Why does Valgrind complain about the use of uninitialized data?

When OpenSSL's PRNG routines are called to generate random numbers the supplied
buffer contents are mixed into the entropy pool: so it technically does not
matter whether the buffer is initialized at this point or not.  Valgrind (and
other test tools) will complain about this. When using Valgrind, make sure the
OpenSSL library has been compiled with the PURIFY macro defined (-DPURIFY)
to get rid of these warnings.
```
